gcloud-java-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java

```diff
@@ -0,0 +1,416 @@
+  public List<Entity> fetch(Iterable<Key> keys, ReadOption... options) {
+    return DatastoreHelper.fetch(this, Iterables.toArray(keys, Key.class), options);
+  }
+
+  Iterator<Entity> get(com.google.datastore.v1beta3.ReadOptions readOptionsPb, final Key... keys) {
+    if (keys.length == 0) {
+      return Collections.emptyIterator();
+    }
+    com.google.datastore.v1beta3.LookupRequest.Builder requestPb =
+        com.google.datastore.v1beta3.LookupRequest.newBuilder();
+    if (readOptionsPb != null) {
+      requestPb.setReadOptions(readOptionsPb);
+    }
+    for (Key k : Sets.newLinkedHashSet(Arrays.asList(keys))) {
+      requestPb.addKeys(k.toPb());
+    }
+    return new ResultsIterator(requestPb);
+  }
+
+  final class ResultsIterator extends AbstractIterator<Entity> {
+
+    private final com.google.datastore.v1beta3.LookupRequest.Builder requestPb;
+    Iterator<com.google.datastore.v1beta3.EntityResult> iter;
+
+    ResultsIterator(com.google.datastore.v1beta3.LookupRequest.Builder requestPb) {
+      this.requestPb = requestPb;
+      loadResults();
+    }
+
+    private void loadResults() {
+      com.google.datastore.v1beta3.LookupResponse responsePb = lookup(requestPb.build());
+      iter = responsePb.getFoundList().iterator();
+      requestPb.clearKeys();
+      if (responsePb.getDeferredCount() > 0) {
+        requestPb.addAllKeys(responsePb.getDeferredList());
+      }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected Entity computeNext() {
+      while (!iter.hasNext()) {
+        if (requestPb.getKeysCount() == 0) {
+          return endOfData();
+        }
+        loadResults();
+      }
+      return Entity.fromPb(iter.next().getEntity());
+    }
+  }
+
+  com.google.datastore.v1beta3.LookupResponse lookup(
+      final com.google.datastore.v1beta3.LookupRequest requestPb) {
+    try {
+      return RetryHelper.runWithRetries(
+          new Callable<com.google.datastore.v1beta3.LookupResponse>() {
+        @Override public com.google.datastore.v1beta3.LookupResponse call()
+            throws DatastoreException {
+          return datastoreRpc.lookup(requestPb);
+        }
+      }, retryParams, EXCEPTION_HANDLER);
+    } catch (RetryHelperException e) {
+      throw DatastoreException.translateAndThrow(e);
+    }
+  }
+
+  @SafeVarargs
+  @Override
+  public final void update(Entity... entities) {
+    if (entities.length > 0) {
+      List<com.google.datastore.v1beta3.Mutation> mutationsPb =
+          new ArrayList<>();
+      Map<Key, Entity> dedupEntities = new LinkedHashMap<>();
+      for (Entity entity : entities) {
+        dedupEntities.put(entity.key(), entity);
+      }
+      for (Entity entity : dedupEntities.values()) {
+        mutationsPb.add(
+            com.google.datastore.v1beta3.Mutation.newBuilder().setUpdate(entity.toPb()).build());
+      }
+      commitMutation(mutationsPb);
+    }
+  }
+
+  @SafeVarargs
+  @Override
+  public final void put(Entity... entities) {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Unnecessary final modifier in final class](https://www.codacy.com/app/mziccard/gcloud-java/file/2397465861/issues/source?bid=3123109&fileBranchId=3231307#l314)